### PR TITLE
Add extension methods to package:collection.

### PR DESCRIPTION
**Pull request by lrhn**
Add a number of useful extension methods to `List` and `Iterable`, and to a few other types.

A good part of these extensions are backed by the algorithms of `algorithms.dart`, which is expanded to support them:

* Added `quickSort`.
* Add extra optional `Random` argument to `shuffle`.
* Generalize some of the functions in `algorithms.dart` to work on specific properties of the objects (`binarySearchBy`,
  `lowerBoundBy`, `insertionSortBy`, `quickSortBy`, `mergeSortBy`).

### CODE REVIEW

**Comment by lrhn** [comment]
This is a first crack at adding extension methods to package:collection.

They are not *tested*. Some are probably not useful (I expect `Iterable<T extends num>.sum()` to be mostly speculative).

File: analysis_options.yaml
Comment (natebosch) [inline analysis_options.yaml:4]: We should be following this lint.

File: analysis_options.yaml
Comment (natebosch) [inline analysis_options.yaml:6]: We should be following this lint.

Some of the removed ones that aren't in `pedantic` yet are also useful. We rarely violate them on the team, but I don't see any reason for something like `camel_case_types` or `library_prefixes` to need to wait for code review if we do get a contribution from outside the team.

File: lib/src/empty_unmodifiable_set.dart
Comment (natebosch) [inline lib/src/empty_unmodifiable_set.dart:11]: This seems like a good change, but I don't understand how it's related. Can this go out in a separate commit?

File: lib/src/algorithms.dart
Comment (natebosch) [inline lib/src/algorithms.dart:32]: I don't think it's a good idea to add new top level methods to this package. Anything new should be added as an extension only. If we want to have these as top level methods internally for some reason, we should be hiding them from our exports.

**Review by natebosch — CHANGES_REQUESTED** [review]
Did something get lost in the upload? There are no extension methods added here. This does look like it might be some work towards adding extensions, but it doesn't match the PR description as is.

File: lib/src/algorithms.dart
Comment (lrhn) [reply lib/src/algorithms.dart:32]: I completely disagree with not adding top-level functions.

Top-level functions are still a perfectly good way to introduce general reusable functionality.

Extension methods are less generally useful than top-level functions, but more convenient and discoverable.
That's why I have made the distinction of adding convenient and discoverable APIs using extension methods, and more rough and verbose reusable APIs as top-level functions.

**Comment by lrhn** [comment]
Whoops. Forgot to git-add the new files.

File: lib/src/empty_unmodifiable_set.dart
Comment (lrhn) [reply lib/src/empty_unmodifiable_set.dart:11]: It's an unrelated drive-by fix. I can do that by itself too.

File: analysis_options.yaml
Comment (lrhn) [reply analysis_options.yaml:6]: As you may  be aware, I disagree with following that lint. It's not part of the Dart style guide, and I actually prefer the inline notation for function arguments, both for reading and writing.
These are the three exceptions that I apply because I consistently write code that does not follow them.

I'm using pedantic+modifications because it's convenient (there are good things in there too, just not all of it), not because I think it's what we actually want as the Dart recommended lint set outside of Google3.
(Sure, I can add camel_case_types, but that has never been an issue.)

Comment (jakemac53) [reply analysis_options.yaml:6]: How is it possibly a good thing to have every random person who touches a codebase changing the lints to fit their personal preference?

Comment (natebosch) [reply analysis_options.yaml:6]: > It's not part of the Dart style guide, and I actually prefer the inline notation for function arguments, both for reading and writing.

https://dart.dev/guides/language/effective-dart/design#consider-using-function-type-syntax-for-parameters

It is a "Consider" it think for largely historical reasons. In future revisions it's likely to become stronger. The fact that there _are_ 2 syntaxes for this will hopefully be fixed in Dart 3 or some 2.x language version - ideally we have 1 way to do things instead of 2 competing ways that require constant negotiation or inconsistent codebases.

> These are the three exceptions that I apply because I consistently write code that does not follow them.

This goes against the flow of the rest of Google Dart authors, and against the general philosophy of Readability in Google.

> Sure, I can add camel_case_types, but that has never been an issue.

It has never been a _big_ issue. It has come up in code reviews in the past and the intention of putting it in a lint is to save time, and save from re-negotiating it with new contributors. The reason `camel_case_types` isn't in `pedantic` is because that would require consistent enforcement, and in internal code I think we have some exceptions. We won't have exceptions in _this_ package which is why the lint is enabled.


I know that we have different approaches to linting in different places. I don't think having us each try to assert our own style into shared packages is the right thing to do, neither is introducing config changes into other PRs.

File: lib/src/empty_unmodifiable_set.dart
Comment (natebosch) [reply lib/src/empty_unmodifiable_set.dart:11]: > I can do that by itself too.

I think that would be better. The travis wait time for individual PRs is the only downside - if you need to send a few things together to work around that I think it's OK, otherwise for isolated refactorings I think we should follow go/small-cls

File: lib/src/algorithms.dart
Comment (natebosch) [reply lib/src/algorithms.dart:32]: > Not every function deserves looking like an instance-method.

When I last looked it seemed like every function in _this_ packages deserves looking like an instance-method. If we wont' be adding `list.binarySearchBy` extension at all, that does change the discussion a bit, but I think that would be a good extension method.

> It also avoids the "all functions after 2.6 are extension methods, all functions before are not" issue.

I don't think that is an issue, I haven't seen anyone make that mistake.

The main thing I want to push on is to have 1 (recommended) way to do things. I want the users of this package to avoid negotiating during _their_ code review about whether to use the extension method version or the top-level method version.

> The general function doesn't really _fit in_ as an instance/extension member, but it's there in case you really need the full power.

I don't understand this. Is this ability to be passed as a function (with a _super_ specific function signature) rather than be torn off a specific list the "full power"?

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:5]: This isn't a useful comment.

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:11]: If the only thing we can say is to repeat the signature, we should omit the comment. Same thing for all the docs on the extensions in this file.

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:21]: Should we match the signature on `List.sort` and make the argument optional?

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:59]: This is inconsistent with `sortedBy` and `sortedCompareBy`. Should we rename this `sortedCompareBy` and add a `sortedBy<K extends Comparable<K>>`?

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:71]: Does this offer enough value over `.toSet()` to make it worth adding?

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:73]: Would `HashSet<T>();` be cheaper? We don't care about maintaining insertion order.

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:207]: I'm not sure if it actually makes a difference (or if we should be trying to micro-optimize without benchmarks) but I _think_ this is faster on dart2js with `??=`

```suggestion
      (result[key] ??= {}).add(element);
```

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:236]: Is this speculative, or has someone asked for this?

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:383]: We need to call out that this is the mean, not the median. Or we could name it `mean` or `arithmeticMean` and avoid the ambiguous `average` term altogether.

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:375]: Should these be getters?

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:483]: The doc should describe the case where `compare` is omitted.

```dart
  /// Whether the element are sorted.
  ///
  /// The order is checked against [compare] if it is passed,
  /// otherwise it is checked against the element's natural order.
```

File: lib/src/list_extensions.dart
Comment (natebosch) [inline lib/src/list_extensions.dart:26]: How much value are we buying by specializing this against `Iterable.mapIndexed`? Do we expect `this[index]` to be meaningfully different (faster?) than `for(var element in this)`?

File: lib/src/list_extensions.dart
Comment (natebosch) [inline lib/src/list_extensions.dart:41]: Typo

```suggestion
  Iterable<R> expandIndexed<R>(Iterable<R> expend(int index, E element)) sync* {
```

File: lib/src/list_extensions.dart
Comment (natebosch) [inline lib/src/list_extensions.dart:49]: I don't think we should add both this _and_ the ability to sort a `ListSlice`

File: lib/src/list_extensions.dart
Comment (natebosch) [inline lib/src/list_extensions.dart:110]: Typo
```suggestion
  /// as this list, and the elements of this list and [other]
```

File: lib/src/list_extensions.dart
Comment (natebosch) [inline lib/src/list_extensions.dart:111]: Comment needs to be rephrased to account for the `equality` argument.

**Review by natebosch — CHANGES_REQUESTED** [review]

File: lib/src/iterable_extensions.dart
Comment (lrhn) [reply lib/src/iterable_extensions.dart:59]: Seems reasonable.
And it should take start/end offsets too.

File: lib/src/iterable_extensions.dart
Comment (lrhn) [reply lib/src/iterable_extensions.dart:71]: Probably not, no.

File: lib/src/list_extensions.dart
Comment (lrhn) [reply lib/src/list_extensions.dart:26]: Yes, but maybe not by a lot.

File: lib/src/iterable_extensions.dart
Comment (lrhn) [reply lib/src/iterable_extensions.dart:21]: The signature of `List.sort` is actually bad. It allows you to omit the `compare`, but that only really works when the elements are `Comparable`. With extension methods we can do *better*.
There is a `sorted([int compare(E a, E b)])` on `Iterable<T extends Comparable<T>>`.

File: analysis_options.yaml
Comment (lrhn) [reply analysis_options.yaml:4]: Done for now.

File: lib/src/algorithms.dart
Comment (lrhn) [reply lib/src/algorithms.dart:32]: There are multiple layers of API here. Both a general low-level API that people can use to build other libraries from, and which isn't necessarily optimized for ease-of-use, and a high-level convenience API that accesses the low-level API in various convenient manners. Actually, most of the former until now. It's a package of building blocks more than a package of solutions.

We could stop exposing the low-level API directly. That would mean not exporting the new quickSort or xBy methods from `algorithms` and only expose them as extension methods. It'll probably work. There will likely be someone asking us to expose it anyway, but if the full power is still accessible as an extension method, it's probably not an issue.

As for `binarySearch`, the reasons I originally didn't make it as an extension is that it doesn't actually work on all lists. It only works on *sorted* lists, and therefore I didn't want it to be an instance method on all lists. (For the same reason I think `sort` is badly designed as well, but that can be saved by extension methods, binary search cannot).
It's ... probably overly cautious of me, and I'll add the binary search as extension methods as well.

Comment (natebosch) [reply lib/src/algorithms.dart]: > There will likely be someone asking us to expose it anyway, but if the full power is still accessible as an extension method, it's probably not an issue.

Waiting to expose something until someone asks is my preferred approach. As a team we've spent a non-trivial amount of time trying to remove stuff that has little or no usage.

**Comment by mit-mit** [comment]
@lrhn looks stale?

**Comment by lrhn** [comment]
PTAL

**Comment by natebosch** [comment]
I had been waiting to review after the merge conflicts were resolved... Do you want me to do a first pass and then re-review after solving the conflicts?

**Comment by lrhn** [comment]
Conflicts should be gone now.

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:26]: Should we make this an optional argument to match `List.sort`?

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:52]: [nit] I think it would be worth calling out the behavior for an empty list in the doc comment.

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:259]: In the SDK I think the `singleWhere` methods throw if there is more than one match, and use `orElse` only for the no-match case. This is a different behavior that may be a better design, but also could be confusing.

If we do want this behavior, can we call out the difference specifically?

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:543]: [nit] Replace all "list" in the comment with "iterable"

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:544]: Should we go with an algorithm that won't overflow?

```dart
extension on Iterable<num> {
  double get average {
    var average = 0.0;
    var count = 1;
    for (var element in this) {
      average += (element - average) / count;
      count++;
    }
    return average;
  }
}
```

File: lib/src/iterable_extensions.dart
Comment (natebosch) [inline lib/src/iterable_extensions.dart:706]: [nit] Why use the function type here and the name `Comparator` below?

File: lib/src/iterable_extensions.dart
Comment (natebosch) [reply lib/src/iterable_extensions.dart:26]: I forgot I had already asked this and we resolved the conversation. Sorry about that!

https://github.com/dart-lang/collection/pull/135#discussion_r425658615

**Review by natebosch — APPROVED** [review]
Overall I'm _slightly_ worried about adding so many of these speculatively since there is a high cost to remove them, but given that they are extension methods I think the overall cost to keeping them is low.

File: CHANGELOG.md
Comment (lrhn) [reply CHANGELOG.md]: Done.  Also removing quickSort reference above, since it's also not exported.

**Review by lrhn — COMMENTED** [review]
Addressed comments.

File: lib/src/algorithms.dart
Comment (lrhn) [reply lib/src/algorithms.dart]: We're now not exposing the new top-level members.

**Pull request merged**
